In [ ]:
import pandas as pd

In [ ]:
df_products = pd.read_csv('uniqlo_product.csv')
df_products.head()

,category,product_url,Product Name,Gender,Size,Price,Rate,Number_of_rates,Description,Chất liệu,Hướng dẫn giặt,Mã sản phẩm,img_url
0,accessories,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,Nữ,One Size,588000,5.0,8,NỮ One Size Túi Dây Rút VND 588.000 New 5.0 (8),Shell: Vải Túi Trong: 100% Polyeste/ Lớp Lót: ...,"Không giặt, Không giặt khô, Không sấy khô",483050,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...
1,accessories,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,Nữ,One Size,588000,4.6,5,NỮ One Size Túi Puffy VND 588.000 New 4.6 (5),Shell: 100% Polyeste/ Lớp Độn: 100% Polyeste/ ...,"Không giặt, Không giặt khô, Không sấy khô",481663,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...
2,bottoms,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,Nữ,XS-XL,588000,4.9,255,NỮ XS-XL Quần Dài Siêu Co Giãn Dáng Rộng VND 5...,"Thân: 77% Nylon, 23% Elastan/ Vải Túi: Lớp Ngo...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483291,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...
3,bottoms,https://www.uniqlo.com/vn/vi/products/E475820-000,Quần Dài Gear,Nữ,M,784000,4.8,111,NỮ M Quần Dài Gear VND 784.000 VND 489.000 Sal...,Thân: 100% Nylon ( 30% Sử Dụng Sợi Nylon Tái C...,"Giặt máy nước lạnh, giặt nhẹ, Giặt khô, Không ...",475820,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...
4,bottoms,https://www.uniqlo.com/vn/vi/products/E472485-000,Quần Jeans Dáng Rộng,Nữ,XS-XXL,980000,4.9,999,NỮ XS-XXL Quần Jeans Dáng Rộng VND 980.000 4.9...,"[00 WHITE] Thân: 99% Bông, 1% Elastan/ Vải Túi...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483813,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...


In [ ]:
df_reviews = pd.read_csv('uniqlo_review.csv')
df_reviews.head()

,product_url,Product Name,Mã sản phẩm,Reviewer,Rating,Comment,Review Date,Gender,Age,Height,Weight,Shoe Size,Country
0,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,483050,Pinkpink,5 out of 5 stars,True to size,15/12/2025,Nữ,25 đến 34 tuổi,141 - 150cm,56 - 60kg,EU36,Penang
1,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Syifa,5 out of 5 stars,Love it,20/12/2025,Nữ,20 đến 24 tuổi,156 - 160cm,46 - 50kg,EU41,Penang
2,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Yong,4 out of 5 stars,More inner pockets!,14/12/2025,Nam,20 đến 24 tuổi,181 - 185cm,71 - 75kg,EU45,Johor
3,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,ahbii,5 out of 5 stars,bag,11/12/2025,Nữ,20 đến 24 tuổi,166 - 170cm,56 - 60kg,EU37,Penang
4,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,483291,Lulu,5 out of 5 stars,Versatile pants for any occasion,13/12/2025,Nữ,35 đến 44 tuổi,156 - 160cm,66 - 70kg,EU39,Kuala Lumpur


In [ ]:
df_products.duplicated().sum()

np.int64(0)

In [ ]:
df_products.columns

Index(['category', 'product_url', 'Product Name', 'Gender', 'Size', 'Price',
       'Rate', 'Number_of_rates', 'Description', 'Chất liệu', 'Hướng dẫn giặt',
       'Mã sản phẩm', 'img_url'],
      dtype='object')

In [ ]:
df_products.duplicated(subset=['product_url', 'Mã sản phẩm']).sum()

np.int64(0)

In [ ]:
df_products.duplicated(subset=['Mã sản phẩm']).sum()

np.int64(3)

In [ ]:
# Identify Product Duplicates (based on ID)
prod_keys = ['Mã sản phẩm']
product_dupes = df_products[df_products.duplicated(subset=prod_keys, keep=False)].sort_values(by=prod_keys)

print("--- Product Duplicates Found ---")
# Showing specific columns to highlight the subtle name differences
product_dupes[['product_url', 'Mã sản phẩm', 'Product Name', 'category']]

--- Product Duplicates Found ---


,product_url,Mã sản phẩm,Product Name,category
283,https://www.uniqlo.com/vn/vi/products/E482449-000,482449,Áo Sơ Mi Oxford Dáng Ôm | Kẻ Sọc,shirts-and-polo-shirts
284,https://www.uniqlo.com/vn/vi/products/E462369-000,482449,Áo Sơ Mi Vải Oxford Dáng Ôm | Kẻ Sọc,shirts-and-polo-shirts
238,https://www.uniqlo.com/vn/vi/products/E482601-000,482601,Quần Lót Boxer Brief | Cạp Thấp,innerwear
239,https://www.uniqlo.com/vn/vi/products/E479502-000,482601,Quần Lót Boxer Briefs Vải Cotton | Cạp Thấp,innerwear
42,https://www.uniqlo.com/vn/vi/products/E485775-000,485775,Áo Khoác Kéo Khóa Dáng Ngắn,outerwear
43,https://www.uniqlo.com/vn/vi/products/E479208-000,485775,Áo Khoác Dáng Ngắn Kéo Khóa,outerwear


In [ ]:
# Identify Review Duplicates (based on User, Product, and Date)
rev_keys = ['Reviewer', 'Mã sản phẩm', 'Review Date']
review_dupes = df_reviews[df_reviews.duplicated(subset=rev_keys, keep=False)].sort_values(by=rev_keys)

print("\n--- Review Duplicates Found ---")
review_dupes[rev_keys + ['Rating']]


--- Review Duplicates Found ---


,Reviewer,Mã sản phẩm,Review Date,Rating
3232,Nabil,482640,27/03/2025,5 out of 5 stars
3233,Nabil,482640,27/03/2025,5 out of 5 stars
76,Trang,483813,04/12/2025,5 out of 5 stars
77,Trang,483813,04/12/2025,5 out of 5 stars


In [ ]:
# 1. Drop duplicates for Product Data based on ID
# We keep 'first' to maintain one version of the product name/metadata
product_df_cleaned = df_products.drop_duplicates(subset=['Mã sản phẩm'], keep='first')

# 2. Drop duplicates for Review Data
# We remove exact duplicates of (User, Product, Date)
review_df_cleaned = df_reviews.drop_duplicates(subset=['Reviewer', 'Mã sản phẩm', 'Review Date'], keep='first')

# 3. Save the "Gold Standard" datasets for ML/GNN training
product_df_cleaned.to_csv('uniqlo_product_cleaned.csv', index=False)
review_df_cleaned.to_csv('uniqlo_review_cleaned.csv', index=False)

In [ ]:
print(df_products.shape)
print(product_df_cleaned.shape)
print(df_reviews.shape)
print(review_df_cleaned.shape)

(352, 13)
(349, 13)
(4457, 13)
(4455, 13)


# Crawl image code

In [ ]:
product_df = pd.read_csv('uniqlo_product_cleaned.csv')
product_df.head()

,category,product_url,Product Name,Gender,Size,Price,Rate,Number_of_rates,Description,Chất liệu,Hướng dẫn giặt,Mã sản phẩm,img_url
0,accessories,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,Nữ,One Size,588000,5.0,8,NỮ One Size Túi Dây Rút VND 588.000 New 5.0 (8),Shell: Vải Túi Trong: 100% Polyeste/ Lớp Lót: ...,"Không giặt, Không giặt khô, Không sấy khô",483050,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...
1,accessories,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,Nữ,One Size,588000,4.6,5,NỮ One Size Túi Puffy VND 588.000 New 4.6 (5),Shell: 100% Polyeste/ Lớp Độn: 100% Polyeste/ ...,"Không giặt, Không giặt khô, Không sấy khô",481663,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...
2,bottoms,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,Nữ,XS-XL,588000,4.9,255,NỮ XS-XL Quần Dài Siêu Co Giãn Dáng Rộng VND 5...,"Thân: 77% Nylon, 23% Elastan/ Vải Túi: Lớp Ngo...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483291,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...
3,bottoms,https://www.uniqlo.com/vn/vi/products/E475820-000,Quần Dài Gear,Nữ,M,784000,4.8,111,NỮ M Quần Dài Gear VND 784.000 VND 489.000 Sal...,Thân: 100% Nylon ( 30% Sử Dụng Sợi Nylon Tái C...,"Giặt máy nước lạnh, giặt nhẹ, Giặt khô, Không ...",475820,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...
4,bottoms,https://www.uniqlo.com/vn/vi/products/E472485-000,Quần Jeans Dáng Rộng,Nữ,XS-XXL,980000,4.9,999,NỮ XS-XXL Quần Jeans Dáng Rộng VND 980.000 4.9...,"[00 WHITE] Thân: 99% Bông, 1% Elastan/ Vải Túi...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483813,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...


In [ ]:
product_df.shape

(349, 13)

In [ ]:
review_df = pd.read_csv('uniqlo_review_cleaned.csv')
review_df.head()

,product_url,Product Name,Mã sản phẩm,Reviewer,Rating,Comment,Review Date,Gender,Age,Height,Weight,Shoe Size,Country
0,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,483050,Pinkpink,5 out of 5 stars,True to size,15/12/2025,Nữ,25 đến 34 tuổi,141 - 150cm,56 - 60kg,EU36,Penang
1,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Syifa,5 out of 5 stars,Love it,20/12/2025,Nữ,20 đến 24 tuổi,156 - 160cm,46 - 50kg,EU41,Penang
2,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Yong,4 out of 5 stars,More inner pockets!,14/12/2025,Nam,20 đến 24 tuổi,181 - 185cm,71 - 75kg,EU45,Johor
3,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,ahbii,5 out of 5 stars,bag,11/12/2025,Nữ,20 đến 24 tuổi,166 - 170cm,56 - 60kg,EU37,Penang
4,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,483291,Lulu,5 out of 5 stars,Versatile pants for any occasion,13/12/2025,Nữ,35 đến 44 tuổi,156 - 160cm,66 - 70kg,EU39,Kuala Lumpur


In [ ]:
review_df.shape

(4455, 13)

In [ ]:
import os
import pandas as pd
import requests
from tqdm import tqdm

# Create the target directory
folder_name = 'product_img'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)
    print(f"Created folder: {folder_name}")

# Download Logic
def download_images(dataframe, folder):
    local_paths = []

    print("Starting download process...")
    for index, row in tqdm(dataframe.iterrows(), total=len(dataframe)):
        url = row['img_url']
        product_id = str(row['Mã sản phẩm'])

        # Define filename and path (using .jpg as seen in URLs)
        filename = f"{product_id}.jpg"
        save_path = os.path.join(folder, filename)

        try:
            # Download image
            response = requests.get(url, stream=True, timeout=10)
            if response.status_code == 200:
                with open(save_path, 'wb') as f:
                    for chunk in response.iter_content(1024):
                        f.write(chunk)
                local_paths.append(filename)
            else:
                local_paths.append("download_failed")
        except Exception as e:
            local_paths.append("error_occurred")

    return local_paths

# Execute and update DataFrame
product_df['product_img'] = download_images(product_df, folder_name)

# 5. Save the updated CSV for your DL models
product_df.to_csv('uniqlo_product_with_images.csv', index=False)
print("Finished! New CSV saved as 'uniqlo_product_with_images.csv'")

Created folder: product_img
Starting download process...


100%|██████████| 349/349 [02:04<00:00,  2.80it/s]

Finished! New CSV saved as 'uniqlo_product_with_images.csv'


In [ ]:
product_df.head()

,category,product_url,Product Name,Gender,Size,Price,Rate,Number_of_rates,Description,Chất liệu,Hướng dẫn giặt,Mã sản phẩm,img_url,product_img
0,accessories,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,Nữ,One Size,588000,5.0,8,NỮ One Size Túi Dây Rút VND 588.000 New 5.0 (8),Shell: Vải Túi Trong: 100% Polyeste/ Lớp Lót: ...,"Không giặt, Không giặt khô, Không sấy khô",483050,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...,483050.jpg
1,accessories,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,Nữ,One Size,588000,4.6,5,NỮ One Size Túi Puffy VND 588.000 New 4.6 (5),Shell: 100% Polyeste/ Lớp Độn: 100% Polyeste/ ...,"Không giặt, Không giặt khô, Không sấy khô",481663,https://image.uniqlo.com/UQ/ST3/AsianCommon/im...,481663.jpg
2,bottoms,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,Nữ,XS-XL,588000,4.9,255,NỮ XS-XL Quần Dài Siêu Co Giãn Dáng Rộng VND 5...,"Thân: 77% Nylon, 23% Elastan/ Vải Túi: Lớp Ngo...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483291,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...,483291.jpg
3,bottoms,https://www.uniqlo.com/vn/vi/products/E475820-000,Quần Dài Gear,Nữ,M,784000,4.8,111,NỮ M Quần Dài Gear VND 784.000 VND 489.000 Sal...,Thân: 100% Nylon ( 30% Sử Dụng Sợi Nylon Tái C...,"Giặt máy nước lạnh, giặt nhẹ, Giặt khô, Không ...",475820,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...,475820.jpg
4,bottoms,https://www.uniqlo.com/vn/vi/products/E472485-000,Quần Jeans Dáng Rộng,Nữ,XS-XXL,980000,4.9,999,NỮ XS-XXL Quần Jeans Dáng Rộng VND 980.000 4.9...,"[00 WHITE] Thân: 99% Bông, 1% Elastan/ Vải Túi...","Giặt máy nước lạnh, giặt nhẹ, Không giặt khô, ...",483813,https://image.uniqlo.com/UQ/ST3/vn/imagesgoods...,483813.jpg


In [ ]:
review_df.head()

,product_url,Product Name,Mã sản phẩm,Reviewer,Rating,Comment,Review Date,Gender,Age,Height,Weight,Shoe Size,Country
0,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,483050,Pinkpink,5 out of 5 stars,True to size,15/12/2025,Nữ,25 đến 34 tuổi,141 - 150cm,56 - 60kg,EU36,Penang
1,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Syifa,5 out of 5 stars,Love it,20/12/2025,Nữ,20 đến 24 tuổi,156 - 160cm,46 - 50kg,EU41,Penang
2,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Yong,4 out of 5 stars,More inner pockets!,14/12/2025,Nam,20 đến 24 tuổi,181 - 185cm,71 - 75kg,EU45,Johor
3,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,ahbii,5 out of 5 stars,bag,11/12/2025,Nữ,20 đến 24 tuổi,166 - 170cm,56 - 60kg,EU37,Penang
4,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,483291,Lulu,5 out of 5 stars,Versatile pants for any occasion,13/12/2025,Nữ,35 đến 44 tuổi,156 - 160cm,66 - 70kg,EU39,Kuala Lumpur


In [ ]:
demographic_groups = review_df.groupby(['Gender', 'Age', 'Weight', 'Height', 'Shoe Size', 'Country']).agg(
    Unique_Reviewers=('Reviewer', 'nunique'),
    Total_Interactions=('Reviewer', 'count')
).reset_index()

# Sort to find the most common customer profiles
demographic_groups = demographic_groups.sort_values(by='Unique_Reviewers', ascending=False)

demographic_groups.shape
# # Save the persona segments
# demographic_groups.to_csv('reviewer_demographics_groups.csv', index=False)

(3293, 8)

- Gender, Age, Height, Shoe Size 682
- Weight 1793
- Country 2381
- All 6 3293

In [ ]:
review_df.shape

(4455, 13)

In [ ]:
demographic_cols = ['Gender', 'Age', 'Height', 'Weight', 'Shoe Size', 'Country']

review_df['reviewer_id'] = review_df.groupby(demographic_cols).ngroup()
review_df.to_csv('uniqlo_review_with_id.csv', index=False)

In [ ]:
review_df.head()

,product_url,Product Name,Mã sản phẩm,Reviewer,Rating,Comment,Review Date,Gender,Age,Height,Weight,Shoe Size,Country,reviewer_id
0,https://www.uniqlo.com/vn/vi/products/E483050-000,Túi Dây Rút,483050,Pinkpink,5 out of 5 stars,True to size,15/12/2025,Nữ,25 đến 34 tuổi,141 - 150cm,56 - 60kg,EU36,Penang,1892
1,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Syifa,5 out of 5 stars,Love it,20/12/2025,Nữ,20 đến 24 tuổi,156 - 160cm,46 - 50kg,EU41,Penang,1708
2,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,Yong,4 out of 5 stars,More inner pockets!,14/12/2025,Nam,20 đến 24 tuổi,181 - 185cm,71 - 75kg,EU45,Johor,337
3,https://www.uniqlo.com/vn/vi/products/E481663-000,Túi Puffy,481663,ahbii,5 out of 5 stars,bag,11/12/2025,Nữ,20 đến 24 tuổi,166 - 170cm,56 - 60kg,EU37,Penang,1828
4,https://www.uniqlo.com/vn/vi/products/E474481-000,Quần Dài Siêu Co Giãn Dáng Rộng,483291,Lulu,5 out of 5 stars,Versatile pants for any occasion,13/12/2025,Nữ,35 đến 44 tuổi,156 - 160cm,66 - 70kg,EU39,Kuala Lumpur,2715
